### PR TITLE
fix(DnsValidatedCertificate): add support for subjectAlternativeNames

### LINKED ISSUE #4659: aws-certificatemanager DnsValidatedCertificateHandler does not properly handle certs with SubjectAlternativeNames
A DnsValidatedCertificate will never successfully validate when SubjectAlternativeNames are present. This seems to be due to the custom resource only adding the first DomainValidationOptions record to Route53. See [line 110](https://github.com/aws/aws-cdk/blob/c0c0513b68cc12e6696c57d2305bf84311ad9c59/packages/%40aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib/index.js#L110) here. This should add a new ResourceRecordSet for every DomainValidationOptions result.

### Reproduction Steps

Create a `DnsValidatedCertificate` and add at least one record to `subjectAlternativeNames`.

### Environment

  - **CLI Version      :** 1.14.0
  - **Framework Version:** 1.14.0
  - **OS               :** MacOS 10.14
  - **Language         :** Typescript

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by touzoku**
- [x] Fix requestCertificate function in packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib/index.js
- [x] Add a test to packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/test/handler.test.js
- [x] Run eslint --fix (sorry for a messy PR, but why the code was not adhering to `.eslintrc.js` rules in the first place?)
- [x] Bumped up node runtime for the certificate requestor Lambda function to `lambda.Runtime.NODEJS_12_X` in packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts

Fixes #4659

----

cc @skinny85 

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts [hunk 1]
```diff
@@ -61,7 +61,7 @@ export class DnsValidatedCertificate extends cdk.Resource implements ICertificat
         const requestorFunction = new lambda.Function(this, 'CertificateRequestorFunction', {
             code: lambda.Code.fromAsset(path.resolve(__dirname, '..', 'lambda-packages', 'dns_validated_certificate_handler', 'lib')),
             handler: 'index.certificateRequestHandler',
-            runtime: lambda.Runtime.NODEJS_10_X,
+            runtime: lambda.Runtime.NODEJS_12_X,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts]: Why this change?

File: packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib/index.js [hunk 2]
```diff
@@ -226,7 +227,7 @@ exports.certificateRequestHandler = async function(event, context) {
           event.ResourceProperties.DomainName,
           event.ResourceProperties.SubjectAlternativeNames,
           event.ResourceProperties.HostedZoneId,
-          event.ResourceProperties.Region,
+          event.ResourceProperties.Region
```
Comment (skinny85) [inline packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib/index.js]: Please do not remove this trailing comma.

File: packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib/index.js [hunk 3]
```diff
@@ -99,42 +99,43 @@ const requestCertificate = async function(requestId, domainName, subjectAlternat
...
+          ResourceRecordSet: {
+            Name: record.Name,
+            Type: record.Type,
+            TTL: 60,
+            ResourceRecords: [{
+              Value: record.Value
+            }]
+          }
+        };
+      })
```
Comment (skinny85) [inline packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib/index.js]: Can you explain why this change in logic is needed in this PR?

File: packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/test/handler.test.js [hunk 4]
```diff
@@ -73,7 +76,7 @@ describe('DNS Validated Certificate Handler', () => {
 
   test('Create operation requests a certificate', () => {
     const requestCertificateFake = sinon.fake.resolves({
-      CertificateArn: testCertificateArn,
+      CertificateArn: testCertificateArn
```
Comment (skinny85) [inline packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/test/handler.test.js]: Please do not remove trailing commas.

File: packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/test/handler.test.js [hunk 5]
```diff
@@ -117,7 +120,7 @@ describe('DNS Validated Certificate Handler', () => {
         ResourceProperties: {
           DomainName: testDomainName,
           HostedZoneId: testHostedZoneId,
-          Region: 'us-east-1',
+          Region: 'us-east-1'
```
Comment (skinny85) [inline packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/test/handler.test.js]: Please do not remove trailing commas.

File: packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/test/handler.test.js [hunk 6]
```diff
@@ -145,10 +148,103 @@ describe('DNS Validated Certificate Handler', () => {
...
+    const describeCertificateFake = sinon.stub();
+    describeCertificateFake.onFirstCall().resolves({
+      Certificate: {
+        CertificateArn: testCertificateArn
+      }
+    });
+    describeCertificateFake.resolves({
+      Certificate: {
+        CertificateArn: testCertificateArn,
+        DomainValidationOptions: [{
```
Comment (skinny85) [inline packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/test/handler.test.js]: Indent this like the following:

```ts
        DomainValidationOptions: [
          {
            ValidationStatus: 'SUCCESS',
            ResourceRecord: {
              Name: testRRName,
              Type: 'CNAME',
              Value: testRRValue,
            },
          },
          {
            ValidationStatus: 'SUCCESS',
            ResourceRecord: {
              Name: testAltRRName,
              Type: 'CNAME',
              Value: testAltRRValue,
            },
          },
        ],
```

**Comment by saltman424** [comment]
> Thanks for the contribution @touzoku !
> 
> But I have to say, I don't follow the change here that well. Can you write a short paragraph explaining what needed changing to support `SubjectAlternativeNames`? From what I can tell, code dealing with them was there already, so I'm not sure what was it that needed changing.

@skinny85 Since I ran into the same issue and had actually also made my own fix, I can give you a quick answer. If you look at lines 110 and 127 - 137 of what is being changed by:
https://github.com/aws/aws-cdk/pull/6516/files#diff-60e5eb022ef3dcfba44e73fe7643fe7f
you can see that at line 110 the only ResourceRecord that was being kept was the one for the first domain, and thus that was the only one for which a CNAME validation record was being upserted in lines 127 - 137. What @touzoku added was logic to wait for all ResourceRecords, and then upsert all required validation records (i.e. the validation record for the first domain as well as all subjectAlternativeNames).

Long story short, the implementation before this pull request only added a single validation record for the domainName, not any of the subjectAlternativeNames. Thus, any DnsValidatedCertificates that specified subjectAlternativeNames would fail to deploy because the certificate would never be completely validated.

File: packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib/index.js (hunk 2 above)
Comment (touzoku) [reply packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib/index.js]: @skinny85 I did not! The eslint config in the package did!

File: packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts (hunk 1 above)
Comment (touzoku) [reply packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts]: Updated nodejs version to the most recent runtime, since this is the only 10.x function in my stacks now. But yes, irrelevant to the PR, side fix.

File: packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/test/handler.test.js (hunk 6 above)
Comment (touzoku) [reply packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/test/handler.test.js]: @skinny85 I have added a rule config to the .eslintrc.json of the lambda-package as follows:
```json
"comma-dangle": ["error", {
    "arrays": "always-multiline",
    "objects": "always-multiline",
     "imports": "never",
    "exports": "never",
    "functions": "never"
}]
```
And ran `eslint --fix` on both `lib/index.js` and `test/handler.test.js`. Let me know if this is what you want. This created a lot of mess, because `test/handler.test.js` did not follow the convention that you have requested that much....

Let me know if this is what you've wanted. Otherwise, tell me what the eslint config should be.

File: packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib/index.js (hunk 3 above)
Comment (touzoku) [reply packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib/index.js]: As @saltman424 mentioned below and as it is pointed out in #4659, current logic only creates a single Route53 validation record and misses creating all other records for subjectAlternativeNames.

File: packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/test/handler.test.js (hunk 6 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/test/handler.test.js]: I'd rather the `.eslintrc.json` file not be modified, and to minimize incidental changes to the lambda code not related to the logic changes.

File: packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts]: Can you bring it back to `NODEJS_10_X` then? I will also accept a property that allows you to set it to `NODEJS_12_X` (but the default should be `NODEJS_10_X`) 🙂.

**Comment by rrrix** [comment]
@skinny85 it looks like this PR will solve #8282 and #7509

I'm thinking perhaps I can fork this PR, make the requested changes (revert `.eslintrc.json`) and make a new PR - unless you know of a way I can update this PR directly? Are there any other changes needed?

**Comment by skinny85** [comment]
> @skinny85 it looks like this PR will solve #8282 and #7509
> 
> I'm thinking perhaps I can fork this PR, make the requested changes (revert `.eslintrc.json`) and make a new PR - unless you know of a way I can update this PR directly? Are there any other changes needed?

Yes, this PR looks quite good, and I really didn't have comments beyond the `.eslintrc.json` changes which make the diff so messy.

Feel free to take it over. Unfortunately, I don't think you can update this PR directly, but we can close it after yours is opened.

**Comment by rrrix** [comment]
I figured out why he changed the `.eslintrc.json` - because your requested modifications (to add trailing commas) conflicts with the existing `.eslintrc.json` :) 

```diff
~/dev/git/aws/aws-cdk/packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib $ eslint index.js

/Users/rbowen/dev/git/aws/aws-cdk/packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib/index.js
   39:25  error  Unexpected trailing comma  comma-dangle
   50:46  error  Unexpected trailing comma  comma-dangle
   51:8   error  Unexpected trailing comma  comma-dangle
   95:28  error  Unexpected trailing comma  comma-dangle
  106:53  error  Unexpected trailing comma  comma-dangle
  133:55  error  Unexpected trailing comma  comma-dangle
  134:12  error  Unexpected trailing comma  comma-dangle
  136:9   error  Unexpected trailing comma  comma-dangle
  138:31  error  Unexpected trailing comma  comma-dangle
  146:22  error  Unexpected trailing comma  comma-dangle
  148:34  error  Unexpected trailing comma  comma-dangle
  156:22  error  Unexpected trailing comma  comma-dangle
  158:51  error  Unexpected trailing comma  comma-dangle
  179:28  error  Unexpected trailing comma  comma-dangle
  202:26  error  Unexpected trailing comma  comma-dangle

✖ 15 problems (15 errors, 0 warnings)
  15 errors and 0 warnings potentially fixable with the `--fix` option.
```

It seems to me the changes @touzoku made in `.eslintrc.json` (which is only for the Lambda function and its tests) are quite reasonable, given your requested formatting changes. 

Shouldn't the `.eslintrc.json` configuration match the desired format? It will probably cause more confusion in the future if they're not aligned.   

Anyways... new PR incoming! 👍 

```diff
diff --git b/packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/.eslintrc.json a/packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/.eslintrc.json
index 12a6e88f4..483a623f6 100644
--- b/packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/.eslintrc.json
+++ a/packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/.eslintrc.json
@@ -1,10 +1,20 @@
 {
     "extends": "standard",
     "rules": {
-        "semi": ["error", "always"]
+        "semi": ["error", "always"],
+        "comma-dangle": ["error", {
+            "arrays": "always-multiline",
+            "objects": "always-multiline",
+            "imports": "never",
+            "exports": "never",
+            "functions": "never"
+        }],
+        "array-bracket-newline": ["error", {
+            "multiline": true
+        }]
     },
     "env": {
         "jest": true,
         "node": true
     }
-}
+}
\ No newline at end of file
```

**Comment by skinny85** [comment]
> I figured out why he changed the `.eslintrc.json` - because your requested modifications (to add trailing commas) conflicts with the existing `.eslintrc.json` :)

Then ignore my comments, and keep the style that passes ESLint :)

**Comment by touzoku** [comment]
I think this PR is obsolete now, somebody else fixed the issue in a different PR.

**Comment by rrrix** [comment]
https://github.com/aws/aws-cdk/pull/6516#issuecomment-642971001

> I think this PR is obsolete now, somebody else fixed the issue in a different PR.

Hi @touzoku! The issue isn't fixed yet (at least, for me). I still get an error when I attempt to create a DNS Validated Certificate with too many subject alternative names. Looking at the latest version of the `master` branch, it looks like the same problematic lambda code that you updated in this PR is still there.  

https://github.com/aws/aws-cdk/blob/86d84e6d592b2bee110ca0fd9890ce32e46055c3/packages/%40aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib/index.js#L108-L119

**Comment by rrrix** [comment]
Looks like CloudFormation natively supports DNS Validation for ACM Certificates now. 

https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/31#issuecomment-643530187 

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-certificatemanager-certificate-domainvalidationoption.html#cfn-certificatemanager-certificate-domainvalidationoption-hostedzoneid

I think perhaps a better PR at this point would be to add the native CFN support to the L2 Construct :) Gets rid of one more pesky CustomResource and Lambda Function.

**Comment by rrrix** [comment]
The "native" (a.k.a. CloudFormation) DNS validated certificate totally worked with SAN's and everything, first try, boom!

Here's how for anyone looking for a workaround instead of the above `DnsValidatedCertificate` custom Lambda fix.

```typescript
import { Certificate } from '@aws-cdk/aws-certificatemanager';
import { HostedZone } from '@aws-cdk/aws-route53';
import { CfnResource, Lazy } from '@aws-cdk/core';

// ... snip

    this.hostedZone = HostedZone.fromLookup(this, 'hostedZone', {
      domainName: this.envDomainName,
    });

    this.cfnCertificate = new CfnResource(this, 'cfnCertificate', {
      type: 'AWS::CertificateManager::Certificate',
      properties: {
        DomainName: this.envDomainName,
        SubjectAlternativeNames: [`*.${this.envDomainName}`],
        ValidationMethod: 'DNS',
        DomainValidationOptions: [
          {
            DomainName: this.envDomainName,
            HostedZoneId: this.hostedZone.hostedZoneId,
          },
        ],
      },
    });

    this.tlsCertificate = Certificate.fromCertificateArn(
      this,
      'certificateArn',
      Lazy.stringValue({ produce: () => this.cfnCertificate.ref }),
    );
```
